tests/test_taar_similarity.py

```diff
@@ -0,0 +1,213 @@
+    # Stub a different set of addons for each variation
+    # so that we get 3 clusters out.
+    CLUSTER_VARIATIONS = [
+        {
+            "count": 50,
+            "variation": {
+                "geo_city": ["Rome"],
+                "subsession_length": [3785],
+                "os": "Linux",
+                "settings": [{
+                    "locale": "it-IT"
+                }],
+            }
+        },
+        {
+            "count": 50,
+            "variation": {
+                "geo_city": ["London"],
+                "subsession_length": [1107],
+                "os": "MacOS",
+                "settings": [{
+                    "locale": "en-UK"
+                }],
+            }
+        },
+        {
+            "count": 50,
+            "variation": {
+                "geo_city": ["Boston"],
+                "subsession_length": [201507],
+                "os": "Windows",
+                "settings": [{
+                    "locale": "en-US"
+                }],
+            }
+        }
+    ]
+
+    sample_snippets = []
+    counter = 0
+    for variation_id, cluster in enumerate(CLUSTER_VARIATIONS):
+        for i in range(0, cluster["count"]):
+            variation = copy.deepcopy(cluster["variation"])
+            variation["client_id"] = "client-{}".format(counter)
+
+            # Generate some valid addons for each client. Each stub
+            # addon as a specific GUID that contains both the variation
+            # id and the addon id for simplified debugging.
+            variation["active_addons"] = [{}]
+            for addon_id in range(0, 3):
+                addon_name = "var-{}-guid-{}".format(variation_id, addon_id)
+                variation["active_addons"][0][addon_name] = {
+                    "blocklisted":     False,
+                    "user_disabled":   False,
+                    "app_disabled":    False,
+                    "signed_state":    2,
+                    "type":            "extension",
+                    "foreign_install": False,
+                    "is_system":       False
+                }
+
+            # Additionally add a system addon.
+            variation["active_addons"][0]["system-addon-guid"] = {
+                "blocklisted":     False,
+                "user_disabled":   False,
+                "app_disabled":    False,
+                "signed_state":    2,
+                "type":            "extension",
+                "foreign_install": False,
+                "is_system":       True
+            }
+
+            sample_snippets.append(variation)
+            counter = counter + 1
+
+    return generate_data(sample_snippets)
+
+
+def test_compute_donors(spark, multi_clusters_df):
+    multi_clusters_df.createOrReplaceTempView("longitudinal")
+
+    # Perform the clustering on our test data. We expect
+    # 3 clusters out of this and 10 donors.
+    donors = taar_similarity.get_donors(spark, 3, 10)
+
+    # assert len(donors) == 10
```

Comment:
There should be an assertion that the number of donors is within a particular margin of the requested number. It will not be equal due to the sampling from variable sized donor candidates (accross clusters) however the number of donors should always be within 2*(n) where n is the number of terminal clusters.

Reply:
Good point. However, to make tests less flaky, I've slightly changed the code to pass around a fixed "random seed" to the Spark ML functions to make tests reproducible. Since we generate the fake data, this allows us to precisely assert how many donors we see.

Reply:
Ah yeah, good point. For generated data it is a deterministic process, given a fixed random seed. Maybe we should add logging in the live module that warns us if this condition is not met so we can assess variability in with the cluster sizes as new client data is pulled in.